dbms/src/Functions/array/range.cpp

```diff
@@ -27,79 +29,144 @@ class FunctionRange : public IFunction
-            return true;
+            out_offsets[row_idx] = offset;
         }
-        else
-            return false;
+
+        block.getByPosition(result).column = ColumnArray::create(std::move(data_col), std::move(offsets_col));
+        return true;
+    }
+
+    template <typename Start, typename End>
+    bool executeStartEnd(Block & block, const IColumn * start_col, const IColumn * end_col, const IColumn * step_col, const size_t result)
+    {
+        return executeStartEndStep<Start, End, UInt8>(block, start_col, end_col, step_col,result)
+               || executeStartEndStep<Start, End, UInt16>(block, start_col, end_col, step_col,result)
+               || executeStartEndStep<Start, End, UInt32>(block, start_col, end_col, step_col,result)
+               || executeStartEndStep<Start, End, UInt64>(block, start_col, end_col, step_col,result);
+    }
+
+    template <typename Start>
+    bool executeStart(Block & block, const IColumn * start_col, const IColumn * end_col, const IColumn * step_col, const size_t result)
+    {
+        return executeStartEnd<Start, UInt8>(block, start_col, end_col, step_col, result)
+               || executeStartEnd<Start, UInt16>(block, start_col, end_col, step_col, result)
+               || executeStartEnd<Start, UInt32>(block, start_col, end_col, step_col, result)
+               || executeStartEnd<Start, UInt64>(block, start_col, end_col, step_col, result);
     }
 
     void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t) override
     {
-        const auto col = block.getByPosition(arguments[0]).column.get();
+        Columns columns_holder(3);
+        ColumnRawPtrs columns(3);
+        size_t idx = 0;
+        size_t rows =  block.getByPosition(arguments[0]).column.get()->size();
 
-        if (!executeInternal<UInt8>(block, col, result) &&
-            !executeInternal<UInt16>(block, col, result) &&
-            !executeInternal<UInt32>(block, col, result) &&
-            !executeInternal<UInt64>(block, col, result))
+        // for start column, default to 0
+        if (arguments.size() == 1)
         {
-            throw Exception{"Illegal column " + col->getName() + " of argument of function " + getName(), ErrorCodes::ILLEGAL_COLUMN};
+            columns_holder[idx] = DataTypeUInt8().createColumnConst(rows, 0)->convertToFullColumnIfConst();
+            columns[idx] = columns_holder[idx].get();
+            idx ++;
+        }
+
+        for (size_t i = 0; i < arguments.size(); ++i)
+        {
+            columns_holder[idx] = block.getByPosition(arguments[i]).column->convertToFullColumnIfConst();
+            columns[idx] = columns_holder[idx].get();
+            idx ++;
+        }
+
+        // for step column, defaults to 1
+        if (arguments.size() <= 2)
+        {
+            columns_holder[idx] = DataTypeUInt8().createColumnConst(rows, 1)->convertToFullColumnIfConst();
```

Comment:
This will slow down the function. Do we have performance test?
(And if we don't care about slowdown, we can also remove cubic template instantiation).